### PR TITLE
Add px guide

### PR DESCRIPTION
**Pull request by sebastianbenz**

### CODE REVIEW

File: packages/page-experience/lib/PageDataGatherer.js [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+      const requestTypeIgnoreList = new Set(['image', 'video']);
+      if (requestTypeIgnoreList.has(request.resourceType())) {
+        return request.abort();
+      }
+      if (
+        request.resourceType() === 'script' &&
+        !request.url().startsWith('https://cdn.ampproject.org')
+      ) {
+        // Only donwload AMP runtime scripts as they're need for layouting the page
+        // Once self-hosting is a thing we'll have to change this
```
Comment (kristoferbaxter) [inline packages/page-experience/lib/PageDataGatherer.js:175]: Is there a flag for when the CLI tools know self-hosting is valid?

File: packages/page-experience/lib/PageExperienceGuide.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    return checksToRun;
+  }
+
+  /**
+   * Loads all  checks from the file system
+   *
+   * @private
+   */
+  async loadChecks() {
+    const jsFileExtension = '.js';
```
Comment (kristoferbaxter) [inline packages/page-experience/lib/PageExperienceGuide.js:127]: Does this work with the recently launched module mode?

File: packages/page-experience/lib/helpers/parseFontface.test.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+test('does not fail on invalid css', () => {
+  const styles = `
+  test test test
+@font-face {
+  font-family: "Open Sans";
+}
+`;
+  parseFontface(styles, 'http://example.com');
+});
```
Comment (kristoferbaxter) [inline packages/page-experience/lib/helpers/parseFontface.test.js:82]: For completeness should there be a check with subset fonts?

File: packages/page-experience/lib/helpers/parseFontfaceSrc.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+const fontfaceSrcParser = require('css-font-face-src');
+
+/**
+ * Extracts the best fontface URL for preloading (woff2 > woff > ?).
```
Comment (kristoferbaxter) [inline packages/page-experience/lib/helpers/parseFontfaceSrc.js:4]: Smart here: I was naively thinking the result would always be `woff2` but this code correctly validates if there are other formats available to use.

File: packages/page-experience/prettier.config.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
```
Comment (kristoferbaxter) [inline packages/page-experience/prettier.config.js:2]: 2021?

**Review by kristoferbaxter — COMMENTED** [review]
Lots of great work here, thank you!

Left you a few nits and other items for the future.

File: packages/page-experience/bin/cli.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const PageExperienceGuide = require('../lib/PageExperienceGuide');
+
+const url = process.argv[2];
+
+(async () => {
+  const result = await new PageExperienceGuide().analyze(url);
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/page-experience/bin/cli.js:22]: Oh yes...

File: packages/page-experience/lib/PageDataGatherer.js [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+const DEFAULT_VIEWPORT = {
+  width: 393,
+  height: 851,
+};
+
+/**
+ * Renders a page in Puppeteer and collects all data required for the page experience recommendations.
+ */
+class PageAnalyzer {
+  constructor(config = {}) {
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/page-experience/lib/PageDataGatherer.js:35]: Added type info

File: packages/page-experience/lib/PageDataGatherer.js [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        return css;
+      };
+
+      /**
+       * Returns a list of all font preload hrefs. URLs are normalized to the current origin.
+       *
+       * @return {Array<string>} a list of URLs
+       */
+      const collectFontPreloads = () => {
+        return Array.from(document.querySelectorAll('link[rel=preload][as=font]')).map(
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/page-experience/lib/PageDataGatherer.js:116]: Not yet, created an issue.

File: packages/page-experience/lib/PageDataGatherer.js [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        document.querySelectorAll('body *').forEach((node) => {
+          const computedStyles = window.getComputedStyle(node);
+          const fontFamily = computedStyles.getPropertyValue('font-family');
+          const font = extractFirstFont(fontFamily);
+          if (!font) {
+            return;
+          }
+          if (isElementInViewport(node)) {
+            criticalFonts.add(font);
+          } else {
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/page-experience/lib/PageDataGatherer.js:165]: Good point, unfortunately not valid in this case. We can't rely on all critical fonts being detected before the non-critical ones. We have to iterate the whole Dom to be able to identify all critical fonts (the last dom element might be in the first viewport).

File: packages/page-experience/lib/helpers/parseFontface.test.js (hunk 3 above)
Comment (sebastianbenz) [reply packages/page-experience/lib/helpers/parseFontface.test.js:85]: done

File: packages/page-experience/prettier.config.js (hunk 5 above)
Comment (sebastianbenz) [reply packages/page-experience/prettier.config.js]: removed the file as no longer needed

File: packages/page-experience/README.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+## Usage
+
+```js
+const PageExperienceGuide = require('@ammproject/toolbox-page-experience');
+
+(async ()=> {
+  const pageExperienceGuide = new PageExperienceGuide();
+  const result = pageExperienceGuide.analyze('https://amp.dev');
+  console.log('result')
+})
```
Comment (patrickkettner) [inline packages/page-experience/README.md:14]: is there a reason this is wrapped in `()` but not invoked?

File: packages/page-experience/lib/PageDataGatherer.js [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+
+  /**
+   * Load a page in Puppeteer and collect data required by checks. Needs to be called after 'start'.
+   *
+   * @param {string} url the URL to analyze
+   * @return {Object} all data collected on the page
+   */
+  async execute(url) {
+    if (!this.started) {
+      throw new Error('Puppeteer not running, please call `start` first.');
```
Comment (patrickkettner) [inline packages/page-experience/lib/PageDataGatherer.js:57]: is there any reason you don't just call start?

File: packages/page-experience/lib/PageDataGatherer.js [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+      if (requestTypeIgnoreList.has(request.resourceType())) {
+        return request.abort();
+      }
+      if (
+        request.resourceType() === 'script' &&
+        !request.url().startsWith('https://cdn.ampproject.org')
+      ) {
+        // Only donwload AMP runtime scripts as they're need for layouting the page
+        // Once self-hosting is a thing we'll have to change this
+        // TODO: investigate whether we could cache these locally
```
Comment (patrickkettner) [inline packages/page-experience/lib/PageDataGatherer.js:217]: You could catch them `https://theheadless.dev/posts/request-interception/`, and reply with local filesystem?

File: packages/page-experience/lib/PageDataGatherer.js [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+        return request.abort();
+      }
+      return request.continue();
+    });
+
+    // Collect external stylesheets from requests as we can't read them otherwise due to CORS
```
Comment (patrickkettner) [inline packages/page-experience/lib/PageDataGatherer.js:223]: Is there a way to miss some stuff as a result of this? Any reason not to just disable CORS in rendering options?

File: packages/page-experience/lib/checks/fontPreloading.js [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+ */
+const checkPreloads = (pageData, result) => {
+  for (const font of pageData.criticalFonts) {
+    const fontface = pageData.fontFaces.get(font);
+    if (!fontface || !fontface.mainSrc) {
+      continue;
+    }
+    if (!pageData.fontPreloads.includes(fontface.mainSrc)) {
+      if (fontface.mainSrc.startsWith('https://fonts.gstatic.com')) {
+        result.warn({
```
Comment (patrickkettner) [inline packages/page-experience/lib/checks/fontPreloading.js:30]: would it make sense to extract these text blobs into an external file?

File: packages/page-experience/lib/checks/fontDisplay.js [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+const checkDisplayOptional = (pageData, result) => {
+  for (const font of pageData.criticalFonts) {
+    const fontface = pageData.fontFaces.get(font);
+    if (!fontface) {
+      continue;
+    }
+
+    if (!fontface.fontDisplay) {
+      result.warn({
+        title: 'Improve LCP using `font-display: optional`',
```
Comment (patrickkettner) [inline packages/page-experience/lib/checks/fontDisplay.js:33]: Would it make sense to extract the warnings into an external file?

**Review by patrickkettner — APPROVED** [review]

**Comment by patrickkettner** [comment]
comments and nits, nothing blocking a LGTM though

File: packages/page-experience/README.md (hunk 10 above)
Comment (sebastianbenz) [reply packages/page-experience/README.md]: not at all...fixed

File: packages/page-experience/lib/PageDataGatherer.js [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const puppeteer = require('puppeteer');
+const treeKill = require('tree-kill');
+const parseFontfaces = require('./helpers/parseFontface');
+
+// Pixel 5 XL
+const DEFAULT_VIEWPORT = {
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/page-experience/lib/PageDataGatherer.js:21]: Afaik no DPI value needed as the default is 1. Not sure why a link would be helpful here.

File: packages/page-experience/lib/PageDataGatherer.js (hunk 11 above)
Comment (sebastianbenz) [reply packages/page-experience/lib/PageDataGatherer.js:64]: You're right, but somehow I preferred the more explicit symmetry of calling start and shutdown. It's an internal API though so I don't really care...

File: packages/page-experience/lib/PageDataGatherer.js (hunk 12 above)
Comment (sebastianbenz) [reply packages/page-experience/lib/PageDataGatherer.js:223]: that was my idea as well

File: packages/page-experience/lib/PageDataGatherer.js (hunk 1 above)
Comment (sebastianbenz) [reply packages/page-experience/lib/PageDataGatherer.js:222]: Not yet, will add later

File: packages/page-experience/lib/PageDataGatherer.js (hunk 13 above)
Comment (sebastianbenz) [reply packages/page-experience/lib/PageDataGatherer.js:229]: That didn't work for me.

File: packages/page-experience/lib/checks/fontDisplay.js (hunk 15 above)
Comment (sebastianbenz) [reply packages/page-experience/lib/checks/fontDisplay.js:33]: For translation?

**Comment by sebastianbenz** [comment]
Thanks for the reviews!

**Pull request merged**